pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	resourceType     = "resource_type"
+	nodeName         = "name"
+	nodeProvisioner  = "provisioner"
+	nodeZone         = "zone"
+	nodeArchitecture = "arch"
+	nodeCapacityType = "capacity_type"
+	nodeInstanceType = "instance_type"
+	nodePhase        = "phase"
+	nodeLabels       = "node_labels"
+)
+
+var (
+	allocatableGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "allocatable",
+			Help:      "Node allocatable",
+		},
+		getLabelNames(),
+	)
+	requestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_requests",
+			Help:      "Node total requests",
+		},
+		getLabelNames(),
+	)
+	limitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_limits",
+			Help:      "Node total limits",
+		},
+		getLabelNames(),
+	)
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_overhead",
+			Help:      "Node total overhead",
+		},
+		getLabelNames(),
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(allocatableGaugeVec)
+	crmetrics.Registry.MustRegister(requestsGaugeVec)
+	crmetrics.Registry.MustRegister(limitsGaugeVec)
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodeZone,
+		nodeArchitecture,
+		nodeCapacityType,
+		nodeInstanceType,
+		nodePhase,
+		nodeLabels,
+	}
+}
+
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName][]prometheus.Labels
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName][]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
```

Comment:
I might not be as onboarded to how GaugeVecs work, but does this mean that we won't retain data on nodes after they're deleted? I think it might be worth distinguishing metrics between long term for user-analysis and short term for capacity optimization. 
 
For example: I can imagine a use-case where someone has a cluster with high churn (e.g. a low value of `ttlSecondsUntilExpired`) on one provisioner, and wants to see how that affects their cost in vs in a provisioner with a higher value of `ttlSecondsUntilExpired`

Reply:
Good question. deleting the GaugeVec does not delete the past data/metrics in Prometheus database. It just tell Prometheus to stop tracking this by closing the active time series. So Prometheus will have the perfect memory of everything that happened in the past.

Reply:
Got it. Does this mean that for some visualization, nodes that don't have datapoints within that time frame won't show up in the legend? If they do, that may be too much clutter.

Reply:
You are exactly right. If there is no data, it will not show up in the graph.